### PR TITLE
Optimization of dynamic_k_matching and bboxes_iou

### PR DESCRIPTION
**Pull request by WayneMao**
COCO train time: 
Before: 1 day, 22:00:54 
After:    1 day, 17:50:37
(64 CPUs, 8GPUs, 64 batch_size, fp16, no cache)

### CODE REVIEW

**Comment by Joker316701882** [comment]
@WayneMao Looks good to me. Please also report your training result for yolox-s here.

File: yolox/models/yolo_head.py [hunk 1]
```diff
@@ -607,26 +607,28 @@ def get_in_boxes_info(
...
         # ---------------------------------------------------------------
-        matching_matrix = torch.zeros_like(cost)
+        device = cost.device
+        matching_matrix = torch.zeros(cost.shape, dtype=torch.uint8, device=device)
 
         ious_in_boxes_matrix = pair_wise_ious
         n_candidate_k = min(10, ious_in_boxes_matrix.size(1))
         topk_ious, _ = torch.topk(ious_in_boxes_matrix, n_candidate_k, dim=1)
         dynamic_ks = torch.clamp(topk_ious.sum(1).int(), min=1)
+        dks = dynamic_ks.tolist()
```
Comment (FateScript) [inline yolox/models/yolo_head.py:617]: no need to  change name, delete  dynamic_ks is called later.

File: yolox/models/yolo_head.py [hunk 2]
```diff
@@ -607,26 +607,28 @@ def get_in_boxes_info(
     def dynamic_k_matching(self, cost, pair_wise_ious, gt_classes, num_gt, fg_mask):
         # Dynamic K
         # ---------------------------------------------------------------
-        matching_matrix = torch.zeros_like(cost)
+        device = cost.device
```
Comment (FateScript) [inline yolox/models/yolo_head.py:610]: what about 
`matching_matrix = torch.zeros_like(cost, dtype=torch.uint8)`

File: yolox/utils/boxes.py [hunk 3]
```diff
@@ -76,30 +76,67 @@ def postprocess(prediction, num_classes, conf_thre=0.7, nms_thre=0.45, class_agn
     return output
 
 
-def bboxes_iou(bboxes_a, bboxes_b, xyxy=True):
+def bboxes_iou(bboxes_a, bboxes_b, xyxy=True, inplace=False):
```
Comment (FateScript) [inline yolox/utils/boxes.py:79]: `inplace = False` in your code is not needed.

File: yolox/utils/boxes.py [hunk 4]
```diff
@@ -76,30 +76,67 @@ def postprocess(prediction, num_classes, conf_thre=0.7, nms_thre=0.45, class_agn
...
     if bboxes_a.shape[1] != 4 or bboxes_b.shape[1] != 4:
         raise IndexError
 
-    if xyxy:
-        tl = torch.max(bboxes_a[:, None, :2], bboxes_b[:, :2])
-        br = torch.min(bboxes_a[:, None, 2:], bboxes_b[:, 2:])
-        area_a = torch.prod(bboxes_a[:, 2:] - bboxes_a[:, :2], 1)
-        area_b = torch.prod(bboxes_b[:, 2:] - bboxes_b[:, :2], 1)
+    if inplace:
```
Comment (FateScript) [inline yolox/utils/boxes.py:83]: I wonder how many gpu memory is saved by this function? @WayneMao

**Comment by FateScript** [comment]
Gentle ping @WayneMao

**Comment by WayneMao** [comment]
> @WayneMao Looks good to me. Please also report your training result for yolox-s here.

image

**Comment by FateScript** [comment]
LGTM. Well done @WayneMao

**Pull request merged**